### PR TITLE
Multi Processing for Setup

### PR DESCRIPTION
**Pull request by rohitmusti**
Hey,

I've been playing around a lot with this codebase wanted to contribute a little bit back. I added in some basic multi-processing and got a nice speed bump for the `process_file` function from `setup.py` without losing accuracy or completeness. It went from taking roughly 78 seconds to process the provided `train-v2.0.json` to roughly 40 seconds to process the same file. I've wanted to experiment with multi-processing for quite some time now and this was the perfect excuse.

The only change it makes is that the ids are slightly out of order because when it splits it up into threads, but since the IDs are self contained and not super super significant, I didn't think it was important to try and maintain them. I did a quick grid search and found 18 to be a decent batch size, but it will definitely vary from machine to machine depending on its resources.

If you pull this in, I definitely recommend squashing commits.

### CODE REVIEW

**Review by chrischute — APPROVED** [review]
Looks great, thanks for the changes

**Comment by rohitmusti** [comment]
Hey so I just noticed an error in my model from when I was training it all weekend and there were no vectors captured during the set up process. Basically, the word and character counters weren't counting correctly when I was running the set up script. I didn't notice the error until I started training. 

In order to make them count correctly, I had to add in some locks into the multi-threading. With the locks in place, the multi-processing functions were actually slower than the regular processing functions. I recommend reverting back to the commit from before this PR was merged. Sorry I didn't test for this, it didn't occur to me that there would be this particular issue. I haven't been able to think of a patch that preserves the speed of the multi-processing. If you have any ideas I'd be happy to try and tinker with them to keep the speed up. Again, sorry I didn't run the training loop end to end before making the PR.

**Comment by chrischute** [comment]
OK. Reverted the commit.
